### ISSUE TITLE
Check for whitespace in CI

### ISSUE THREAD

**Issue by erlend-aasland** [python/core-workflow#429]
Consider adding a CI check for whitespace errors, a la `git diff --check`.

I've seen this idea surface a couple of times, most recently by @JelleZijlstra in https://github.com/python/cpython/pull/31695#issuecomment-1059782285

**Comment by erlend-aasland** [comment]
I remembered I got a PoC branch lying around: https://github.com/erlend-aasland/cpython/tree/add-whitespace-linter

**Comment by hugovk** [comment]
Sounds good.

Looking at the https://github.com/erlend-aasland/cpython/pull/14 PoC, some ideas:

* Rather than only running on PRs, can it also run for pushes? That way contributors may see and fix a failure on their fork's CI even before opening a PR ("fast feedback loop")

* How about using https://pre-commit.com ? There's a [`trailing-whitespace` hook](https://pre-commit.com/hooks) to trim trailing whitespace.

  * For developers who like pre-commit, it can autofix things for you before you even commit/push

  * For those who don't, there's absolutely no obligation to run locally and the CI will pick things up

  * Can run on pushes and PRs 

  * Easily extensible, there are lots of other checks which may be useful (for example, `check-yaml` to validate CI config)

**Comment by erlend-aasland** [comment]
> Rather than only running on PRs, can it also run for pushes? That way contributors may see and fix a failure on their fork's CI even before opening a PR ("fast feedback loop")

+1

> How about using https://pre-commit.com/ ?

Oooh, that sure looks nice!

**Comment by hugovk** [comment]
Here's an example: https://github.com/hugovk/cpython/pull/22

There's currently a lot of files with trimmable trailing whitespace, see:

* https://github.com/hugovk/cpython/pull/22#issuecomment-1059804459 for a list 
* https://github.com/hugovk/cpython/runs/5434180892?check_suite_focus=true for diffs

So we could limit it to certain file types. For example `types: [c, python]`. And/or to certain directories.

**Comment by erlend-aasland** [comment]
> Here's an example: [hugovk/cpython#22](https://github.com/hugovk/cpython/pull/22)

Neat! Much more elegant than my PoC 😅 

> So we could limit it to certain file types. For example `types: [c, python]`. And/or to certain directories.

+1

**Comment by hugovk** [comment]
I've updated https://github.com/hugovk/cpython/pull/22 to remove the trailing whitespace, and add:

```yml
        types_or: [c, python]
        exclude: ^Lib/test/test_argparse.py$
```

I excluded that file because a couple of its asserts depend on trailing whitespace:

https://github.com/python/cpython/blob/6927632492cbad86a250aa006c1847e03b03e70b/Lib/test/test_argparse.py#L2179
https://github.com/python/cpython/blob/6927632492cbad86a250aa006c1847e03b03e70b/Lib/test/test_argparse.py#L2196

**Comment by arhadthedev** [comment]
> I excluded that file because a couple of its asserts depend on trailing whitespace

These whitespaces are invisible to a programmer who reads the assertions. Can the following be used instead?

```diff
         self.assertEqual(parser.format_help(), textwrap.dedent('''\
             usage: PROG [-h] {}
             main description
             positional arguments:
-              {}          
+              {}          ''' + '''
             options:
               -h, --help  show this help message and exit
         '''))
```

**Comment by ericvsmith** [comment]
> > I excluded that file because a couple of its asserts depend on trailing whitespace
> 
> These whitespaces are invisible to a programmer who reads the assertions. Can the following be used instead?

I agree that's better. Also a comment explaining why there are trailing spaces and why the string is constructed that way would help.

**Comment by hugovk** [comment]
Well turns out there was already a whitespace check run on the CI as part of `make patchcheck`.

It was broken but has now been fixed: https://github.com/python/cpython/pull/31708

Do we still need the pre-commit suggestion in https://github.com/hugovk/cpython/pull/22 or this issue?

**Comment by sweeneyde** [comment]
I think it would be nice to have the patchcheck as its own GitHub Action with its own green checkmark, so that sort of common issue gets noticed within seconds, rather than having to wait on the build and the whole test suite to run.

Theoretically, I think we should be able to use `actions/setup-python@v3` and not even have to build python at all. Currently, patchcheck uses `sys.version` to determine what the base branch is, but it could be determined by some sort of `patchcheck.py --base_branch ${{ github.base_ref }}` feature instead. But I am no wizard at GitHub Actions.

**Comment by hugovk** [comment]
Please see PR https://github.com/python/cpython/pull/104275.

**Comment by erlend-aasland** [comment]
See also:

- python/cpython#53158
